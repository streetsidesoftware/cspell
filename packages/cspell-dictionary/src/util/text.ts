const regExFirstUpper = /^\p{Lu}\p{M}?\p{Ll}+$/u;
const regExAllUpper = /^(?:\p{Lu}\p{M}?)+$/u;
const regExAllLower = /^(?:\p{Ll}\p{M}?)+$/u;
const regExAccents = /\p{M}/gu;

export function isUpperCase(word: string): boolean {
    return !!word.match(regExAllUpper);
}

export function isLowerCase(word: string): boolean {
    return !!word.match(regExAllLower);
}

export function isFirstCharacterUpper(word: string): boolean {
    return isUpperCase(word.slice(0, 1));
}

export function isFirstCharacterLower(word: string): boolean {
    return isLowerCase(word.slice(0, 1));
}

export function ucFirst(word: string): string {
    return word.slice(0, 1).toUpperCase() + word.slice(1);
}

export function lcFirst(word: string): string {
    return word.slice(0, 1).toLowerCase() + word.slice(1);
}

export function matchCase(example: string, word: string): string {
    if (example.match(regExFirstUpper)) {
        return word.slice(0, 1).toUpperCase() + word.slice(1).toLowerCase();
    }
    if (example.match(regExAllLower)) {
        return word.toLowerCase();
    }
    if (example.match(regExAllUpper)) {
        return word.toUpperCase();
    }

    if (isFirstCharacterUpper(example)) {
        return ucFirst(word);
    }

    if (isFirstCharacterLower(example)) {
        return lcFirst(word);
    }

    return word;
}

export function removeAccents(text: string): string {
    return text.normalize('NFD').replace(regExAccents, '');
}

export function removeUnboundAccents(text: string): string {
    return text.replace(regExAccents, '');
}
