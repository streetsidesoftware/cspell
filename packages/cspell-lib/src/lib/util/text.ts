import { opConcatMap, opMap, pipe } from '@cspell/cspell-pipe/sync';
import type { TextDocumentOffset, TextOffset } from '@cspell/cspell-types';

import type { Uri } from './IUri.js';
import { binarySearch } from './search.js';
import {
    regExAccents,
    regExAllLower,
    regExAllUpper,
    regExFirstUpper,
    regExIgnoreCharacters,
    regExpCamelCaseWordBreaksWithEnglishSuffix,
    regExWords,
    regExWordsAndDigits,
} from './textRegex.js';
import { toUri } from './Uri.js';
import { scanMap } from './util.js';

export { stringToRegExp } from './textRegex.js';

// CSpell:ignore ings ning gimuy tsmerge

export function splitCamelCaseWordWithOffset(wo: TextOffset): TextOffset[] {
    return splitCamelCaseWord(wo.text).map(
        scanMap<string, TextOffset>((last, text) => ({ text, offset: last.offset + last.text.length }), {
            text: '',
            offset: wo.offset,
        }),
    );
}

/**
 * Split camelCase words into an array of strings.
 */
export function splitCamelCaseWord(word: string): string[] {
    return splitWord(word, regExpCamelCaseWordBreaksWithEnglishSuffix);
}

export function splitWordWithOffset(wo: TextOffset, regExpWordBreaks: RegExp): TextOffset[] {
    return splitWord(wo.text, regExpWordBreaks).map(
        scanMap<string, TextOffset>((last, text) => ({ text, offset: last.offset + last.text.length }), {
            text: '',
            offset: wo.offset,
        }),
    );
}

/**
 * Split camelCase words into an array of strings.
 */
export function splitWord(word: string, regExpWordBreaks: RegExp): string[] {
    return word.split(regExpWordBreaks);
}

/**
 * This function lets you iterate over regular expression matches.
 */
export function match(reg: RegExp, text: string): Iterable<RegExpExecArray> {
    if (!text) return [];
    reg = reg.global ? reg : new RegExp(reg.source, reg.flags + 'g');
    return text.matchAll(reg);
}

export function matchStringToTextOffset(reg: RegExp, text: string): Iterable<TextOffset> {
    return matchToTextOffset(reg, { text, offset: 0 });
}

export function matchToTextOffset(reg: RegExp, t: TextOffset): Iterable<TextOffset> {
    const text = t.text;
    const offset = t.offset;
    // return opMap((m: RegExpExecArray) => ({ text: m[0], offset: offset + m.index }))(match(reg, text));
    return pipe(
        match(reg, text),
        opMap((m) => ({ text: m[0], offset: offset + m.index })),
    );
}

export function* extractLinesOfText(text: string): Iterable<TextOffset> {
    let i = 0;
    for (let j = text.indexOf('\n', i); j >= 0; j = text.indexOf('\n', i)) {
        const end = j + 1;
        yield { text: text.slice(i, end), offset: i };
        i = end;
    }
    yield { text: text.slice(i, text.length), offset: i };
}

/**
 * Extract out whole words from a string of text.
 */
export function extractWordsFromText(text: string): Iterable<TextOffset> {
    return extractWordsFromTextOffset(textOffset(text));
}

/**
 * Extract out whole words from a string of text.
 */
export function extractWordsFromTextOffset(text: TextOffset): Iterable<TextOffset> {
    const reg = new RegExp(regExWords);
    return matchToTextOffset(reg, cleanTextOffset(text));
}

/**
 * Remove Hiragana, Han, Katakana, Hangul characters from the text.
 * @param text
 * @returns the text with the characters removed.
 */
export function cleanText(text: string): string {
    regExIgnoreCharacters.lastIndex = 0;
    if (!regExIgnoreCharacters.test(text)) return text;
    text = text.replace(regExIgnoreCharacters, (match: string) => ' '.repeat(match.length));
    return text;
}

export function cleanTextOffset(text: TextOffset): TextOffset {
    // Do not make a new object if the text is already clean.
    regExIgnoreCharacters.lastIndex = 0;
    if (!regExIgnoreCharacters.test(text.text)) return text;
    return {
        text: cleanText(text.text),
        offset: text.offset,
    };
}

/**
 * Extract out whole words and words containing numbers from a string of text.
 */
export function extractPossibleWordsFromTextOffset(text: TextOffset): Iterable<TextOffset> {
    const reg = new RegExp(regExWordsAndDigits);
    return matchToTextOffset(reg, text);
}

export function extractWordsFromCode(text: string): Iterable<TextOffset> {
    return extractWordsFromCodeTextOffset(textOffset(text));
}

export function extractWordsFromCodeTextOffset(textOffset: TextOffset): Iterable<TextOffset> {
    return pipe(extractWordsFromTextOffset(textOffset), opConcatMap(splitCamelCaseWordWithOffset));
}

export function isUpperCase(word: string): boolean {
    return regExAllUpper.test(word);
}

export function isLowerCase(word: string): boolean {
    return regExAllLower.test(word);
}

export function isFirstCharacterUpper(word: string): boolean {
    return isUpperCase(word.slice(0, 1));
}

export function isFirstCharacterLower(word: string): boolean {
    return isLowerCase(word.slice(0, 1));
}

export function ucFirst(word: string): string {
    return word.slice(0, 1).toUpperCase() + word.slice(1);
}

export function lcFirst(word: string): string {
    return word.slice(0, 1).toLowerCase() + word.slice(1);
}

export function snakeToCamel(word: string): string {
    return word.split('_').map(ucFirst).join('');
}

export function camelToSnake(word: string): string {
    return splitCamelCaseWord(word).join('_').toLowerCase();
}

export function matchCase(example: string, word: string): string {
    if (regExFirstUpper.test(example)) {
        return word.slice(0, 1).toUpperCase() + word.slice(1).toLowerCase();
    }
    if (regExAllLower.test(example)) {
        return word.toLowerCase();
    }
    if (regExAllUpper.test(example)) {
        return word.toUpperCase();
    }

    if (isFirstCharacterUpper(example)) {
        return ucFirst(word);
    }

    if (isFirstCharacterLower(example)) {
        return lcFirst(word);
    }

    return word;
}

export function textOffset(text: string, offset = 0): TextOffset {
    return { text, offset };
}

export function extractText(textOffset: TextOffset, startPos: number, endPos: number): string {
    const { text, offset: orig } = textOffset;
    const a = Math.max(startPos - orig, 0);
    const b = Math.max(endPos - orig, 0);
    return text.slice(a, b);
}

export function calculateTextDocumentOffsets<T extends TextOffset>(
    uri: string | Uri | URL,
    doc: string,
    wordOffsets: T[],
): (TextDocumentOffset & T)[] {
    const lines = [
        -1,
        ...pipe(
            match(/\n/g, doc),
            opMap((a) => a.index),
        ),
        doc.length,
    ];

    let lastRow = -1;
    let lastOffset = doc.length + 1;
    let lastLineRow = -1;
    let lastLine: TextOffset | undefined;

    function findRowCol(offset: number): [number, number] {
        const row = binarySearch(lines, offset, offset >= lastOffset ? lastRow : undefined);
        const col = offset - lines[Math.max(0, row - 1)];
        lastOffset = offset;
        lastRow = row;
        return [row, col];
    }

    function extractLine(row: number): TextOffset {
        const offset = lines[row - 1] + 1;
        const text = doc.slice(offset, lines[row] + 1);
        return { text, offset };
    }

    function calcLine(row: number): TextOffset {
        const last = lastLineRow === row ? lastLine : undefined;
        lastLineRow = row;
        const r = last ?? extractLine(row);
        lastLine = r;
        return r;
    }

    const _uri = toUri(uri).toString();

    return wordOffsets.map((wo) => {
        const [row, col] = findRowCol(wo.offset);
        return { ...wo, row, col, doc, uri: _uri, line: calcLine(row) };
    });
}

export function removeAccents(text: string): string {
    return text.normalize('NFD').replace(regExAccents, '');
}

export const __testing__ = {
    regExWords,
    regExWordsAndDigits,
};
