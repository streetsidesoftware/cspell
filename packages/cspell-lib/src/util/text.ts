import { xregexp as XRegExp } from 'cspell-util-bundle';
import {Sequence, sequenceFromRegExpMatch } from 'gensequence';
import {binarySearch} from './search';
import { scanMap } from './util';

// CSpell:ignore ings ning gimuy tsmerge

export interface TextOffset {
    text: string;
    offset: number;
}

export interface TextDocumentOffset extends TextOffset {
    uri?: string;
    doc: string;
    row: number;
    col: number;
}

const regExLines = /.*(\r?\n|$)/g;
// const regExIdentifiers = XRegExp('(?:\\p{L}|[0-9_\'])+', 'gi');
const regExUpperSOrIng = XRegExp('(\\p{Lu}+\\\\?[\'’]?(?:s|ing|ies|es|ings|ed|ning))(?!\\p{Ll})', 'g');
const regExSplitWords = XRegExp('(\\p{Ll})(\\p{Lu})', 'g');
const regExSplitWords2 = XRegExp('(\\p{Lu})(\\p{Lu}\\p{Ll})', 'g');
const regExWords = XRegExp("\\p{L}(?:\\\\?['’]\\p{L}|\\p{L})+|\\p{L}", 'g');
const regExIgnoreCharacters = XRegExp('\\p{Hiragana}|\\p{Han}|\\p{Katakana}|[\\u30A0-\\u30FF]|[\\p{Hangul}]', 'g');
const regExFirstUpper = XRegExp('^\\p{Lu}\\p{Ll}+$');
const regExAllUpper = XRegExp('^\\p{Lu}+$');
const regExAllLower = XRegExp('^\\p{Ll}+$');

const regExMatchRegExParts = /^\/(.*)\/([gimuy]*)$/;

const regExAccents = XRegExp('\\p{M}', 'g');

export function splitCamelCaseWordWithOffset(wo: TextOffset): Array<TextOffset> {
    return splitCamelCaseWord(wo.text)
        .map(scanMap<string, TextOffset>(
            (last, text) => ({ text, offset: last.offset + last.text.length }),
            { text: '', offset: wo.offset }
        ));
}

/**
 * Split camelCase words into an array of strings.
 */
export function splitCamelCaseWord(word: string): string[] {
    const wPrime = word.replace(regExUpperSOrIng, s => s[0] + s.substr(1).toLowerCase());
    const separator = '_<^*_*^>_';
    const pass1 = XRegExp.replace(wPrime, regExSplitWords, '$1' + separator + '$2');
    const pass2 = XRegExp.replace(pass1, regExSplitWords2, '$1' + separator + '$2');
    return XRegExp.split(pass2, separator);
}

/**
 * This function lets you iterate over regular expression matches.
 */
export function match(reg: RegExp, text: string): Sequence<RegExpExecArray> {
    return sequenceFromRegExpMatch(reg, text);
}

export function matchStringToTextOffset(reg: RegExp, text: string) {
    return matchToTextOffset(reg, { text, offset: 0 });
}

export function matchToTextOffset(reg: RegExp, text: TextOffset): Sequence<TextOffset> {
    const textOffset = text;
    const fnOffsetMap = offsetMap(textOffset.offset);
    return match(reg, textOffset.text)
        .map(m => fnOffsetMap({ text: m[0], offset: m.index }));
}

export function extractLinesOfText(text: string): Sequence<TextOffset> {
    return matchStringToTextOffset(regExLines, text);
}

/**
 * Extract out whole words from a string of text.
 */
export function extractWordsFromText(text: string): Sequence<TextOffset> {
    return extractWordsFromTextOffset(textOffset(text));
}

/**
 * Extract out whole words from a string of text.
 */
export function extractWordsFromTextOffset(text: TextOffset): Sequence<TextOffset> {
    const reg = XRegExp(regExWords);
    const reg2 = XRegExp(regExWords);
    return matchToTextOffset(reg, text)
        // remove characters that match against \p{L} but are not letters (Chinese characters are an example).
        .map(({ text, offset }) => ({
            text: XRegExp.replace(text, regExIgnoreCharacters, match => ' '.repeat(match.length)),
            offset,
        }))
        .concatMap(wo => matchToTextOffset(reg2, wo))
        .filter(wo => !!wo.text);
}

export function extractWordsFromCode(text: string): Sequence<TextOffset> {
    return extractWordsFromCodeTextOffset(textOffset(text));
}

export function extractWordsFromCodeTextOffset(textOffset: TextOffset): Sequence<TextOffset> {
    return extractWordsFromTextOffset(textOffset)
        .concatMap(splitCamelCaseWordWithOffset);
}

export function isUpperCase(word: string) {
    return !!word.match(regExAllUpper);
}

export function isLowerCase(word: string) {
    return !!word.match(regExAllLower);
}

export function isFirstCharacterUpper(word: string) {
    return isUpperCase(word.slice(0, 1));
}

export function isFirstCharacterLower(word: string) {
    return isLowerCase(word.slice(0, 1));
}

export function ucFirst(word: string) {
    return word.slice(0, 1).toUpperCase() + word.slice(1);
}

export function lcFirst(word: string) {
    return word.slice(0, 1).toLowerCase() + word.slice(1);
}

export function snakeToCamel(word: string) {
    return word.split('_').map(ucFirst).join('');
}

export function camelToSnake(word: string) {
    return splitCamelCaseWord(word).join('_').toLowerCase();
}

export function matchCase(example: string, word: string): string {
    if (example.match(regExFirstUpper)) {
        return word.slice(0, 1).toUpperCase() + word.slice(1).toLowerCase();
    }
    if (example.match(regExAllLower)) {
        return word.toLowerCase();
    }
    if (example.match(regExAllUpper)) {
        return word.toUpperCase();
    }

    if (isFirstCharacterUpper(example)) {
        return ucFirst(word);
    }

    if (isFirstCharacterLower(example)) {
        return lcFirst(word);
    }

    return word;
}

export function textOffset(text: string, offset: number = 0): TextOffset {
    return { text, offset };
}

export function extractText(textOffset: TextOffset, startPos: number, endPos: number) {
    const { text, offset: orig } = textOffset;
    const a = Math.max(startPos - orig, 0);
    const b = Math.max(endPos - orig, 0);
    return text.slice(a, b);
}

interface OffsetMap {
    offset: number;
}
function offsetMap(offset: number) {
    return <T extends OffsetMap>(xo: T) => ({...(xo as Object), offset: xo.offset + offset }) as T;
}

export function stringToRegExp(pattern: string | RegExp, defaultFlags = 'gim', forceFlags = 'g') {
    if (pattern instanceof RegExp) {
        return pattern;
    }
    try {
        const [, pat, flag] = [...(pattern.match(regExMatchRegExParts) || ['', pattern, defaultFlags]), forceFlags];
        // Make sure the flags are unique.
        const flags = [...(new Set(forceFlags + flag))].join('').replace(/[^gimuy]/g, '');
        if (pat) {
            const regex = new RegExp(pat, flags);
            return regex;
        }
    } catch (e) {
    }
    return undefined;
}

export function calculateTextDocumentOffsets(uri: string, doc: string, wordOffsets: TextOffset[]): TextDocumentOffset[] {
    const lines = [-1, ...match(/\n/g, doc).map(a => a.index), doc.length];

    function findRowCol(offset: number): [number, number] {
        const row = binarySearch(lines, offset);
        const col = offset - lines[Math.max(0, row - 1)];
        return [row, col];
    }

    return wordOffsets
        .map(wo => {
            const [row, col] = findRowCol(wo.offset);
            return { ...wo, row, col, doc, uri };
        });
}

export function removeAccents(text: string) {
    return text.normalize('NFKD').replace(regExAccents, '');
}
