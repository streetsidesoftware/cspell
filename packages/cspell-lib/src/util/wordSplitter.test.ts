import { TextOffset } from './text';
import { SortedBreaks, split, __testing__ } from './wordSplitter';

const generateWordBreaks = __testing__.generateWordBreaks;

const words = sampleWordSet();
const regHasLetters = /\p{L}/u;

describe('Validate wordSplitter', () => {
    interface TestApplyWordBreaks {
        text: string;
        expected: string[];
    }

    test('expensive split', () => {
        // cspell:disable
        // The following is an example of a very expensive split
        const randomText = 'token = "Usf3uVQOZ9m6uPfVonKR-EBXjPe7bjMbp3_Fq8MfsptgkkM1ojidN0BxYaT5HAEN1";';
        // cspell:enable

        const line = {
            text: randomText,
            offset: 0,
        };
        const result = split(line, 9, has);
        const words = result.words.map((w) => w.text).join('|');
        // cspell:ignore VQOZ Mfsptgkk ojid HAEN
        expect(words).toBe('Usf|u|VQOZ|m|u|Pf|Von|KR|EB|Xj|Pe|bj|Mbp|Fq|Mfsptgkk|M|ojid|N|Bx|Ya|T|HAEN');
        expect(result.words.filter((w) => !w.isFound).length).toBe(7);
    });

    // cspell:ignore MOVSX
    test.each`
        text                | expected
        ${'hello'}          | ${['hello']}
        ${'well-educated'}  | ${['well', 'educated']}
        ${'ERRORCode'}      | ${['ERROR', 'Code']}
        ${'MOVSX_r_rm16'}   | ${['MOVSX', 'r', 'rm']}
        ${'32bit-checksum'} | ${['bit', 'checksum']}
        ${'camelCase'}      | ${['camel', 'Case']}
    `('Extract word breaks to $text', ({ text, expected }: TestApplyWordBreaks) => {
        const line = {
            text,
            offset: 42,
        };
        const lineSeg = {
            line: line,
            relStart: 0,
            relEnd: text.length,
        };
        const posBreaks = generateWordBreaks(lineSeg, {});

        const breaks = extractBreaks(posBreaks);
        const r = applyWordBreaks(line, breaks);
        expect(r.map((t) => t.text)).toEqual(expected);
    });

    test.each`
        text                | expected
        ${'hello'}          | ${['hello']}
        ${'well-educated'}  | ${['well|educated', 'well|-educated', 'well-|educated', 'well-educated']}
        ${'MOVSX_r_rm16'}   | ${['MOVSX|r|rm']}
        ${'32bit-checksum'} | ${['bit|checksum']}
        ${'ERRORCode'}      | ${['ERROR|Code']}
        ${'camelCase'}      | ${['camel|Case', 'camelCase']}
    `('Extract all possible word breaks to $text', ({ text, expected }: TestApplyWordBreaks) => {
        const line = {
            text,
            offset: 42,
        };
        const lineSeg = {
            line: line,
            relStart: 0,
            relEnd: text.length,
        };

        const posBreaks = generateWordBreaks(lineSeg, {});
        const r = genAllPossibleResults(text, posBreaks);
        expect(r[0]).toEqual(expected[0]); // Expect the first candidate to be the one with the shortest words.
        expect(r).toEqual(expect.arrayContaining(expected));
        expect(r).toEqual(expect.arrayContaining([text])); // this assumes the original text is a candidate
        expect(r).toMatchSnapshot(); // Use snapshots to ensure all possible options are generated.
    });

    interface PartialTextOffsetWithIsFound {
        text: string;
        offset?: number;
        isFound?: boolean;
    }

    interface TestSplit {
        text: string;
        expectedWords: PartialTextOffsetWithIsFound[];
    }

    /** to PartialTextOffsetWithIsFound */
    function tov(p: PartialTextOffsetWithIsFound | string, defaultIsFound = true): PartialTextOffsetWithIsFound {
        if (typeof p === 'string') {
            p = { text: p };
        }
        const { isFound = defaultIsFound } = p;
        return { ...p, isFound };
    }

    function splitTov(t: string): PartialTextOffsetWithIsFound[] {
        if (!t) return [];
        const parts = t.split('|');
        return parts.map((p) => tov(p, has({ text: p, offset: 0 })));
    }

    // cspell:ignore CVTPD CVTSI CVTTSD words'separated'by errorcode
    // cspell:word Geschäft gescha
    test.each`
        text                                | expectedWords
        ${'hello'}                          | ${[tov({ text: 'hello', offset: 155 })]}
        ${'well-educated'}                  | ${[tov('well'), tov('educated')]}
        ${'MOVSX_r_rm16'}                   | ${splitTov('MOVSX_r_rm16')}
        ${'32bit-checksum'}                 | ${splitTov('bit|checksum')}
        ${'ERRORCode'}                      | ${splitTov('ERROR|Code')}
        ${'camelCase'}                      | ${splitTov('camel|Case')}
        ${'CVTPD2PS_x_xm'}                  | ${splitTov('CVTPD2PS|x|xm')}
        ${'CVTSI2SD_x_rm'}                  | ${splitTov('CVTSI|SD|x|rm')}
        ${'CVTTSD2SI_r_xm'}                 | ${splitTov('CVTTSD|SI|r|xm')}
        ${'error_code42_one_two'}           | ${splitTov('error|code42|one|two')}
        ${'_errorcode42_one_two'}           | ${splitTov('_errorcode42|one|two')}
        ${"words'separated'by_singleQuote"} | ${splitTov(`words|separated|by|singleQuote`)}
        ${"Tom's_hardware"}                 | ${splitTov("Tom's|hardware")}
        ${'Geschäft'}                       | ${splitTov('Geschäft')}
    `('split $text', ({ text, expectedWords }: TestSplit) => {
        const prefix = 'this is some';
        const line = {
            text: `${prefix} ${text} to split.`,
            offset: 142,
        };
        const offset = line.offset + prefix.length;
        const r = split(line, offset, has);
        expect(r.offset).toBe(offset);
        expect(r.endOffset).toBe(r.text.offset + r.text.text.length);
        expect(r.words).toEqual(expect.arrayContaining(expectedWords.map(expect.objectContaining)));
        expect(r.words).toHaveLength(expectedWords.length);
    });

    interface TestSplitWithCalls extends TestSplit {
        calls: number;
    }

    test.each`
        text                                | expectedWords                                 | calls
        ${'hello'}                          | ${[tov({ text: 'hello', offset: 142 })]}      | ${1}
        ${''}                               | ${[]}                                         | ${0}
        ${'#@()&*'}                         | ${[]}                                         | ${0}
        ${'well-educated'}                  | ${[tov('well'), tov('educated')]}             | ${2}
        ${'MOVSX_r_rm16'}                   | ${splitTov('MOVSX_r_rm16')}                   | ${8}
        ${'32bit-checksum'}                 | ${splitTov('bit|checksum')}                   | ${2}
        ${'ERRORCodesTwo'}                  | ${splitTov('ERROR|Codes|Two')}                | ${4}
        ${'camelCase'}                      | ${splitTov('camel|Case')}                     | ${2}
        ${'CVTPD2PS_x_xm'}                  | ${splitTov('CVTPD2PS|x|xm')}                  | ${6}
        ${'CVTSI2SD_x_rm'}                  | ${splitTov('CVTSI|SD|x|rm')}                  | ${10}
        ${'errCVTTSD2SI_r_xm'}              | ${splitTov('err|CVTTSD|SI|r|xm')}             | ${12}
        ${"words'separated'by_singleQuote"} | ${splitTov('words|separated|by|singleQuote')} | ${6}
        ${"Tom's_hardware"}                 | ${splitTov("Tom's|hardware")}                 | ${5}
    `('split edge cases `$text`', ({ text, expectedWords, calls }: TestSplitWithCalls) => {
        const line = {
            text,
            offset: 142,
        };
        const offset = line.offset;
        const h = jest.fn();
        const hasCalls: string[] = [];
        h.mockImplementation((t) => {
            hasCalls.push(t.text);
            return has(t);
        });
        const r = split(line, offset, h, { optionalWordBreakCharacters: `'` });
        // console.log(hasCalls);
        expect(r.offset).toBe(offset);
        expect(r.endOffset).toBe(r.text.offset + r.text.text.length);
        expect(r.words).toEqual(expect.arrayContaining(expectedWords.map(expect.objectContaining)));
        expect(r.words).toHaveLength(expectedWords.length);
        expect(h).toHaveBeenCalledTimes(calls);
    });

    interface TestSplit2 {
        text: string;
        expectedWords: string;
        calls: number;
    }

    // cspell:ignore nstatic techo n'cpp n'log refactor'd î
    test.each`
        text              | expectedWords      | calls
        ${'static'}       | ${'static'}        | ${1}
        ${'nstatic'}      | ${'static'}        | ${1}
        ${'techo'}        | ${'echo'}          | ${1}
        ${`n'cpp`}        | ${'cpp'}           | ${1}
        ${`î'cpp`}        | ${'î|cpp'}         | ${2}
        ${`îphoneStatic`} | ${'îphone|Static'} | ${2}
        ${`êphoneStatic`} | ${'êphone|Static'} | ${2}
        ${`geschäft`}     | ${'geschäft'}      | ${1}
        ${`n'log`}        | ${'log'}           | ${9}
        ${'64-bit'}       | ${'bit'}           | ${1}
        ${'128-bit'}      | ${'bit'}           | ${1}
        ${'256-sha'}      | ${'256-sha'}       | ${6}
        ${`REFACTOR'd`}   | ${'REFACTOR'}      | ${3}
        ${`dogs'`}        | ${`dogs'`}         | ${2}
        ${`planets’`}     | ${`planets’`}      | ${2}
    `('split `$text` in doc', ({ text, expectedWords, calls }: TestSplit2) => {
        const expectedWordSegments = splitTov(expectedWords);
        const doc = sampleText();
        const line = findLine(doc, text);
        const offset = line.offset + line.text.indexOf(text);
        expect(offset).toBeGreaterThan(0);
        const h = jest.fn();
        const hasCalls: string[] = [];
        h.mockImplementation((t) => {
            hasCalls.push(t.text);
            return has(t);
        });
        const r = split(line, offset, h);
        // console.log(hasCalls);
        expect(r.endOffset).toBe(r.text.offset + r.text.text.length);
        expect(r.words).toEqual(expect.arrayContaining(expectedWordSegments.map(expect.objectContaining)));
        expect(r.words).toHaveLength(expectedWordSegments.length);
        expect(h).toHaveBeenCalledTimes(calls);
    });
});

function has({ text }: TextOffset): boolean {
    const nfcText = text.normalize('NFC');
    return text.length < 3 || !regHasLetters.test(text) || words.has(nfcText) || words.has(nfcText.toLowerCase());
}

function applyWordBreaks(text: TextOffset, breaks: number[]): TextOffset[] {
    const a = text.offset;
    const t = text.text;
    const words: TextOffset[] = [];
    let i = 0;
    for (let p = 0; p < breaks.length; p += 2) {
        const start = breaks[p];
        const end = breaks[p + 1];
        if (i !== start) {
            words.push({
                text: t.slice(i, start),
                offset: a + i,
            });
        }
        i = end;
    }

    if (i < t.length) {
        words.push({
            text: t.slice(i),
            offset: a + i,
        });
    }

    return words;
}

function extractBreaks(pwb: SortedBreaks): number[] {
    const r: number[] = [];
    for (const b of pwb) {
        const br = b.breaks[0];
        if (br) {
            r.splice(r.length, 0, ...br);
        }
    }
    return r;
}

function genAllPossibleResults(text: string, breaks: SortedBreaks): string[] {
    function* genResults(i: number, bi: number): Iterable<string> {
        const br = breaks[bi];
        if (!br) {
            yield text.slice(i);
            return;
        }

        const emitted = new Set<string>();
        for (const b of br.breaks) {
            const parts: string[] = [];
            let p = i;
            for (let x = 0; x < b.length; x += 2) {
                const s = b[x];
                const e = b[x + 1];
                parts.push(text.slice(p, s));
                p = e;
            }
            const prefix = parts.join('|');
            const prefix2 = prefix + (prefix ? '|' : '');
            for (const r of genResults(p, bi + 1)) {
                const t = r ? prefix2 + r : prefix;
                if (!emitted.has(t)) {
                    emitted.add(t);
                    yield t;
                }
            }
        }
    }

    return [...genResults(0, 0)];
}

function findLine(doc: string, text: string): TextOffset {
    const index = doc.indexOf(text);
    let lastLine: RegExpMatchArray | undefined = undefined;
    for (const line of doc.matchAll(/.*/g)) {
        if ((line.index || 0) < index) {
            lastLine = line;
        }
    }
    if (lastLine && lastLine.index) {
        return {
            text: lastLine[0],
            offset: lastLine.index,
        };
    }
    return {
        text: '',
        offset: -1,
    };
}

function sampleWordSet() {
    const words = `
    _errorcode42
    2SD
    64-bit
    bit checksum
    camel case
    can't
    code42
    const
    cpp
    CVTPD2PS
    CVTTSD
    echo
    îphone
    êphone
    Geschäft
    error codes
    hello
    MOVSX_r_rm16
    one two
    static
    these are some sample words
    Tom's hardware
    well educated
    words separated by singleQuote
    256-sha
    dogs'
    leashes
    writers
    planets’
    `
        .split(/\s+/g)
        .map((a) => a.trim())
        .filter((a) => !!a);
    return new Set(words);
}

function sampleText() {
    return `
    static const char new_stub1_1[] = "\\techo"
    \\n'log' => 'text/plain'
	static const char new_stub2[] = "';\\nconst LEN = ";
    static const char new_stub3_0[] = ";\\n\\nstatic function go($return = false)\\n'cpp'"

    /* The escape was a back (or forward) reference. We keep the offset in
    order to give a more useful diagnostic for a bad forward reference. For
    references to groups numbered less than 10 we can't use more than two items
    in parsed_pattern because they may be just two characters in the input (and
    in a 64-bit world an offset may need two elements). So for them, the offset
    ERROR's
    REFACTOR'd
    of the first occurrent is held in a special vector. */

    256-sha

    - The dogs' leashes (multiple dogs).
    - The writers' desks (multiple writers).
    - The planets’ atmospheres (multiple planets).

    128-bit values

    î'cpp
    îphoneStatic

    geschäft

    êphoneStatic

`;
}

// cspell:ignore êphone îphone geschäft
