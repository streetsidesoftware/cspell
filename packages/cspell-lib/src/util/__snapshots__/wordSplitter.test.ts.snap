// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate wordSplitter Extract all possible word breaks to 32bit-checksum 1`] = `
[
  "bit|checksum",
  "bit|-checksum",
  "bit-|checksum",
  "bit-checksum",
  "32bit|checksum",
  "32bit|-checksum",
  "32bit-|checksum",
  "32bit-checksum",
  "32|bit|checksum",
  "32|bit|-checksum",
  "32|bit-|checksum",
  "32|bit-checksum",
]
`;

exports[`Validate wordSplitter Extract all possible word breaks to ERRORCode 1`] = `
[
  "ERROR|Code",
  "ERRORC|ode",
  "ERRORCode",
]
`;

exports[`Validate wordSplitter Extract all possible word breaks to MOVSX_r_rm16 1`] = `
[
  "MOVSX|r|rm",
  "MOVSX|r|rm|16",
  "MOVSX|r|rm16",
  "MOVSX|r|_rm",
  "MOVSX|r|_rm|16",
  "MOVSX|r|_rm16",
  "MOVSX|r_|rm",
  "MOVSX|r_|rm|16",
  "MOVSX|r_|rm16",
  "MOVSX|r_rm",
  "MOVSX|r_rm|16",
  "MOVSX|r_rm16",
  "MOVSX|_r|rm",
  "MOVSX|_r|rm|16",
  "MOVSX|_r|rm16",
  "MOVSX|_r|_rm",
  "MOVSX|_r|_rm|16",
  "MOVSX|_r|_rm16",
  "MOVSX|_r_|rm",
  "MOVSX|_r_|rm|16",
  "MOVSX|_r_|rm16",
  "MOVSX|_r_rm",
  "MOVSX|_r_rm|16",
  "MOVSX|_r_rm16",
  "MOVSX_|r|rm",
  "MOVSX_|r|rm|16",
  "MOVSX_|r|rm16",
  "MOVSX_|r|_rm",
  "MOVSX_|r|_rm|16",
  "MOVSX_|r|_rm16",
  "MOVSX_|r_|rm",
  "MOVSX_|r_|rm|16",
  "MOVSX_|r_|rm16",
  "MOVSX_|r_rm",
  "MOVSX_|r_rm|16",
  "MOVSX_|r_rm16",
  "MOVSX_r|rm",
  "MOVSX_r|rm|16",
  "MOVSX_r|rm16",
  "MOVSX_r|_rm",
  "MOVSX_r|_rm|16",
  "MOVSX_r|_rm16",
  "MOVSX_r_|rm",
  "MOVSX_r_|rm|16",
  "MOVSX_r_|rm16",
  "MOVSX_r_rm",
  "MOVSX_r_rm|16",
  "MOVSX_r_rm16",
]
`;

exports[`Validate wordSplitter Extract all possible word breaks to camelCase 1`] = `
[
  "camel|Case",
  "camelCase",
]
`;

exports[`Validate wordSplitter Extract all possible word breaks to hello 1`] = `
[
  "hello",
]
`;

exports[`Validate wordSplitter Extract all possible word breaks to well-educated 1`] = `
[
  "well|educated",
  "well|-educated",
  "well-|educated",
  "well-educated",
]
`;
